src/apps/lwaftr/fragmentv4.lua

```diff
@@ -0,0 +1,100 @@
+module(..., package.seeall)
+
+local constants = require("apps.lwaftr.constants")
+local packet = require("core.packet")
+local ffi = require("ffi")
+
+local band, bor, lshift = bit.band, bit.bor, bit.lshift
+local ntohs, htons = ffi.C.ntohs, ffi.C.htons
+
+local int16_ptr_t = ffi.typeof("uint16_t*")
+
+local total_length_offset = constants.ethernet_header_size + constants.o_ipv4_total_length
+local frag_id_offset = constants.ethernet_header_size + constants.o_ipv4_identification
+local flags_and_frag_offset_offset = constants.ethernet_header_size + constants.o_ipv4_flags
+local ihl_and_ver_offset = constants.ethernet_header_size + constants.o_ipv4_ver_and_ihl
+
+-- Constants to manipulate the flags next to the frag-offset field directly
+-- as a 16-bit integer, without needing to shift the 3 flag bits.
+local flag_dont_fragment_mask  = 0x4000
+local flag_more_fragments_mask = 0x2000
+local frag_offset_field_mask   = 0x1FFF
+
+
+local function get_u16_at_offset(pkt, offset)
+	return ntohs(ffi.cast(int16_ptr_t, pkt.data + offset)[0])
+end
+
+local function set_u16_at_offset(pkt, offset, value)
+	ffi.cast(int16_ptr_t, pkt.data + offset)[0] = htons(value)
+end
+
+-- TODO: Consider security/performance tradeoffs of randomization
+local internal_frag_id = 0x4242
+local function fresh_frag_id()
+	internal_frag_id = band(internal_frag_id + 1, 0xFFFF)
+	return internal_frag_id
+end
+
+-- IPv4 fragmentation, as per https://tools.ietf.org/html/rfc791
+function fragment_ipv4(ipv4_pkt, mtu)
+	if ipv4_pkt.length <= mtu then
+		return ipv4_pkt -- No fragmentation needed
+	end
+
+	-- Discard packets with the DF (dont't fragment) flag set
+	-- TODO: Check the RFCs, maybe this should return an ICMP error
+	if band(get_u16_at_offset(ipv4_pkt, flags_and_frag_offset_offset), flag_dont_fragment_mask) ~= 0 then
+		return nil
+	end
+
+	-- Given as the amount of 32-bit words, in the lower nibble
+	local ihl = band(ipv4_pkt.data[ihl_and_ver_offset], 0xF) * 4
```

Comment:
There's a function in lwaftr.lua to retrieve the `ihl` field of an IPv4 packet. I think, to avoid code duplication, we should try to refactor this code to make it reusable.